File: frontend/src/model/product.ts
// 商品主體
export default class Product {
    id: number;
    name: string;
    price: {
        min: number;
        max: number;
    };
    specTypes: SpecType[];              // 最多兩種
    variants: ProductVariant[];         // 所有的規格組合

    constructor(
        id: number,
        name: string,
        specTypes: SpecType[],
        variants: ProductVariant[]
    ) {
        this.id = id;
        this.name = name;
        this.specTypes = specTypes;
        this.variants = variants;

        // 自動計算 minPrice 和 maxPrice
        const prices = variants.map((variant) => variant.price);
        this.price = {
            min: Math.min(...prices), // 最小價格
            max: Math.max(...prices), // 最大價格
        };
    }
}

// 規格類型（最多兩種：例如顏色、尺寸）
export interface SpecType {
    name: string;                     // 規格名稱（例如：顏色）
    values: SpecValue[];             // 對應的值
}

// 規格值（第一個規格需要圖片）
export interface SpecValue {
    value: string;                   // 規格值（例如：紅色）
    imageUrl?: string;               // 第一種規格才會填入圖片
}

// 商品的規格組合（每一筆子商品資料）
export interface ProductVariant {
    specValue1: string;
    specValue2?: string;
    stock: number;
    price: number;
}

File: backend/models/product.go
// models/product.go
package models

import "gorm.io/gorm"

type Product struct {
	ID         uint           `gorm:"primaryKey"`
	Name       string         `json:"name"`
	PriceMin   float64        `json:"priceMin"`
	PriceMax   float64        `json:"priceMax"`
	SpecTypes  []SpecType     `gorm:"foreignKey:ProductID" json:"specTypes"`
	Variants   []ProductVariant `gorm:"foreignKey:ProductID" json:"variants"`
	CreatedAt  int64
	UpdatedAt  int64
}

type SpecType struct {
	gorm.Model
	ProductID uint         `json:"-"`
	Name      string       `json:"name"`
	Values    []SpecValue  `gorm:"foreignKey:SpecTypeID" json:"values"`
}

type SpecValue struct {
	gorm.Model
	SpecTypeID uint   `json:"-"`
	Value      string `json:"value"`
	ImageURL   string `json:"imageUrl,omitempty"` // 只有第一種規格會填
}

type ProductVariant struct {
	gorm.Model
	ProductID   uint    `json:"-"`
	SpecValue1  string  `json:"specValue1"`
	SpecValue2  string  `json:"specValue2,omitempty"`
	Stock       int     `json:"stock"`
	Price       float64 `json:"price"`
}

func CalculatePriceRange(variants []ProductVariant) (min, max float64) {
	if len(variants) == 0 {
		return 0, 0
	}
	min, max = variants[0].Price, variants[0].Price
	for _, v := range variants {
		if v.Price < min {
			min = v.Price
		}
		if v.Price > max {
			max = v.Price
		}
	}
	return
}